src/Api/PubnubApi/EventEngine/EffectDispatcher.cs

```diff
@@ -0,0 +1,72 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace PubnubApi.PubnubEventEngine
+{
+	public class EffectDispatcher
+	{
+		public enum DispatcherType
+		{
+			Entry,
+			Exit,
+			Managed
+		}
+		public IPubnubUnitTest PubnubUnitTest { get; set; }
+
+		public Dictionary<EventType, IEffectInvocationHandler> effectActionMap;
+		public EffectDispatcher()
+		{
+			effectActionMap = new Dictionary<EventType, IEffectInvocationHandler>();
+		}
+
+		public async void dispatch(EventType effect, ExtendedState stateContext)
+		{
+			IEffectInvocationHandler? handler;
+			if (effectActionMap.TryGetValue(effect, out handler)) {
+				if (handler != null)
+				{
+					await Task.Factory.StartNew(()=> handler.Start(stateContext, effect)).ConfigureAwait(false);
+				}
+			}
+		}
+
+		public async void dispatch(DispatcherType dispatchType, EventType eventType,List<EffectInvocation> effectInvocations, ExtendedState stateContext)
```

Comment:
```
		public async void dispatch(List<EffectInvocationType> invocations, ExtendedState stateContext)
		{
			
			IEffectHandler handler;
			foreach (var invocation in effectInvocations) {
				if(invocation.isManaged() && effectActionMap.TryGetValue(effect, out handler)){
					currentOngoingEffect = handler;
				await Task.Run(()=> handler.Start(stateContext));	
			}
				else if (invocation.isCancelling()){
					if(currentOngoingEffect!= null){
						currentOngoingEffect.cancel()
					}
				}
				else {
					if(effectActionMap.TryGetValue(effect, out handler)){
						handler.start(stateContext);		// non managed oneshot effect.. simple method execution
					}
				}
			}
		}
```

In this way, Effect dispatcher doesn't need to know about state's entry/exit.
Dispatcher will only manage the effect identification from invocation and it's management

Reply:
sure..we can refactor this way.